File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max = Integer.MIN_VALUE;
    public int maxNodeSum(TreeNode root)
    {
        if(root == null)    return 0;

        int l = maxNodeSum(root.left);
        int r = maxNodeSum(root.right);
        System.out.println(root.val + " " + l + " " + r);
        max = Math.max(max, (root.val+ Math.max(0,  l) + Math.max(0,  r)));
        System.out.println("Max : " + max);
        
        int ret = Math.max(l, r);
        ret = Math.max(0, ret);
        return ret+root.val;
    }
    public int maxPathSum(TreeNode root) 
    {
        maxNodeSum(root);
        return max;
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int height(TreeNode* root, int &sum){
        if(root == nullptr) return 0;
        int left = max(0, height(root->left, sum));
        int right = max(0, height(root->right, sum));
        sum = max(sum, left+right+root->val);
        return root->val + max(left, right);
    }
    int maxPathSum(TreeNode* root) {
        int sum = INT32_MIN;
        height(root, sum);
        return sum;
    }
};